sw/device/lib/base/status.h

```diff
@@ -0,0 +1,135 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_STATUS_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_STATUS_H_
+#include <assert.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/macros.h"
+
+#define USING_ABSL_STATUS
+#include "sw/device/lib/base/absl_status.h"
+#undef USING_ABSL_STATUS
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * We use the error category codes from absl_status.h.  We build a packed
+ * status value that identifies the source of the error (in the form of the
+ * module identifier and line number).
+ *
+ * By default, the module identifier is the first three letters of the
+ * source filename.  The identifier can be overridden (per-module) with the
+ * DECLARE_MODULE_ID macro.
+ *
+ * Our status codes are arranged as a packed bitfield, with the sign
+ * bit signifying whether the value represents a result or an error.
+ *
+ * All Ok (good) values:
+ * 32  31                                             0
+ *  +---+---------------------------------------------+
+ *  |   |                  31 bit                     |
+ *  | 0 |                  Result                     |
+ *  +---+---------------------------------------------+
+ *
+ * All Error values:
+ * 32  31      26      21      16             5       0
+ *  +---+-------+-------+-------+-------------+-------+
+ *  |   |   15 bit              | 11 bit      | 5 bit |
+ *  | 1 |   Module Identifier   | Line Number | code  |
```

Comment:
nit: Did you get the chance to check what is the max number of lines for our repo? I'd expect to be significantly less than 2047, but still wanted to check :)

Reply:
The great majority of our source files (`*.{c,h,cc}` are 100 lines or fewer.  Currently, the longest is `spi_device_unittest.cc` coming in at 1850 lines.  Of the 761 non-vendor source files, only 10 are longer than 1000 lines.

The only files longer than 2048 lines are the vendored-in freertos_kernel files.